xonsh/built_ins.py

```diff
@@ -45,7 +45,13 @@ def __init__(self, *args, **kwargs):
         if len(args) == 0 and len(kwargs) == 0:
             args = (os.environ,)
         for key, val in dict(*args, **kwargs).items():
-            self[key] = val
+            if is_function_string(val):
+                if key in (ENV or {}):
```

Comment:
It is much faster (and more explicit) for this to be `if ENV is None or key in ENV`